### PR TITLE
Embed device keys in Olm-encrypted messages

### PR DESCRIPTION
**Pull request by uhoreg**
- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

**Comment by uhoreg** [comment]
A couple issues with the current approach:

- we don't seem to have the cross-signed device key available, so we need to re-cross-sign it every time, but that's an async operation, so it makes a bunch of things async.  Is there a better way?
- the function for cross-signing is only available within the crypto crate, but I'd rather do the signing in the storage crates so that we only need to do the signing once (and avoid some complexity), but I'm not sure what are the implications of changing the function from `pub(crate)`  to `pub`

**Comment by poljar** [comment]
> * we don't seem to have the cross-signed device key available, so we need to re-cross-sign it every time, but that's an async operation, so it makes a bunch of things async.  Is there a better way?

We do, well we do if a `/keys/query` has been done. It's part of the `Device` in the store. You'll have the same async problem though, since the `Device` isn't cached. I think re-cross-signing the device keys is not a good idea. We should use this signature iff it has been published to the server and it has to be the exact same signature that was published to the server, otherwise we'll just confuse ourselves. We'll have multiple valid, albeit different signatures for the same device keys.

> * the function for cross-signing is only available within the crypto crate, but I'd rather do the signing in the storage crates so that we only need to do the signing once (and avoid some complexity), but I'm not sure what are the implications of changing the function from `pub(crate)`  to `pub`

I think that letting stores cross-sign should not happen at all, we don't control all store implementations and this is an obvious shift of responsiblity where store implementations need to care about cryptography to get things working correctly.

**Comment by uhoreg** [comment]
I think that this is ready to be reviewed now.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/account.rs:911]: Please can we have an entry in the above doc comment for this param?

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/account.rs:982]: Describe in the doc comment please.

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/session.rs:237]: Is this a FIXME we should do before merging?

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/session.rs:320]: This would be slightly easier to read with the `let ... else` pattern, which was recently added https://doc.rust-lang.org/beta/rust-by-example/flow_control/let_else.html

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/session.rs:293]: I'm finding it hard to work out what this test does. Could it have a more specific name, and/or could you add some comments? I often add "given, when, then" comments to explain what I'm doing, but here what I'm really looking for is which line of code is the actual thing we're testing?

Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/session.rs:293]: Is it "Given some stuff, when we create an inbound session, then it contains the signed device keys"?

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/session_manager/sessions.rs:519]: Could we actually fetch the device keys up front to avoid starting off with this empty and filling it in later? It might make this easier to read but I assume there's a good reason we're doing it lazily?

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [inline crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:453]: We need some some documentation of the return type here.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [inline crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:547]: Should this be `!=`? I don't fully understand the code, but if we're looking for a changed device ID I'd expect to be looking for a different value here.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [inline crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:546]: Do we need the `clone` here? I'd expect to be able to just inspect the value so could do it by reference.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [inline crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:731]: Should this be inside an `else`? I'd expect to want to add the sessions to the cache even if we've just cleared the cache. I expect I'm misunderstanding what this is doing.

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (andybalaam) [inline crates/matrix-sdk-sqlite/src/crypto_store.rs:815]: Same question about `!=`

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:547]: Ah, no - it's a new _key_ for the _same_ device, right?

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (andybalaam) [inline crates/matrix-sdk-sqlite/src/crypto_store.rs:1413]: Is there some test we can add to `cryptostore_integration_tests` for this? Then it would cover indexeddb, which I think is not covered so far in this PR.

**Review by andybalaam — COMMENTED** [review]
This looks good, with some minor comments and questions scattered around.

My main question is: if we don't find the device keys in the store, we generate them. Shouldn't we save them to the store at that point, so we don't end up generating them multiple times and risking inconsistency as @poljar mentioned in his previous comment?

I don't feel confident to give this full approval so will ask @poljar to take a look.

**Comment by poljar** [comment]
> My main question is: if we don't find the device keys in the store, we generate them.

Have not looked at the whole PR yet, but please take a look at our constructor:

https://github.com/matrix-org/matrix-rust-sdk/blob/abda95979691c4d51d6b7fc06a9777d9e32a795d/crates/matrix-sdk-crypto/src/machine.rs#L297-L311

We are guaranteed to have our own device keys and device in the store. We can add a method to the store called `get_own_device()` which will always return a device to make this clearer. Of course, this does not mean that we're guaranteed to have the cross-signing keys attached to our device.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (uhoreg) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:546]: We need the `clone` because `is_some_and` takes over the `account_info`, and we're in a `for` loop. Though, taking another look at it, replacing `clone` with `as_ref` seems to make the borrow checker happy, and looks like a better thing to do.

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:519]: If there are no sessions to save, then we don't need to fetch the device keys.  So it saves us from doing work if it isn't needed.  I don't know how much work it is to fetch the device keys, so I don't know if it makes a difference.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (uhoreg) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:547]: Yup, we only need to clear the cache if we get new device keys for our own device (e.g. if we get new cross-signing signatures).

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (uhoreg) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:731]: We don't know if the sessions were created using the old or the new device keys, so we don't want to cache them.  (In fact, they're probably created using the old device keys, since we hadn't saved the new device keys yet).

**Comment by uhoreg** [comment]
@andybalaam I think that I've addressed your comments.  There's some CI issues that I still need to look into, but it should be reviewable otherwise.

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/session.rs:237]: Right, I meant to put a comment about that in the PR.  I'm wondering if we should change the return value to a `Result`, or if we should just do an `unwrap` or `expect` and panic if the keys aren't there.  We should be pretty safe for the time being, but I'm a bit worried that if we change crypto systems in the future, this will come back and bite us.  I would lean towards just doing an `unwrap`/`expect`, but I wanted a second opinion.

Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/session.rs:237]: When in doubt, prefer `Result`, I'd say.

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs
Comment (andybalaam) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs:735]: Makes sense - please could you add a comment explaining this?

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/session.rs:184]: Is this `unwrap` OK? If so, please can we have a comment explaining why?

**Review by andybalaam — COMMENTED** [review]
Looks good from my point of view, when the not generating device key part is done. Thanks!

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (uhoreg) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs:1397]: I don't think we can add this to `cryptostore_integration_tests` because `memorystore` works differently, and always keeps the sessions in cache (the cache can't be cleared). (But yes, I should add a test for indexeddb -- once I figure out how to run the indexeddb tests locally)

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs]: With the new function for getting our device from storage, I decided to just fetch the keys every time -- we're already fetching the account every time.

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/session.rs:184]: It's safe, if [the FIXME](https://github.com/matrix-org/matrix-rust-sdk/pull/3517#discussion_r1636123693) gets fixed.  I'll wait for @poljar's review before changing anything here.

**Comment by uhoreg** [comment]
@poljar I think this is ready for review from you now. Some notes

- It currently just uses the device keys from storage, and assumes that it will get updated after cross-signing is done, to include the cross-signing signatures. Is this true, or is there anything else that needs to be done? Is there a test that I should write to make sure that it gets updated?
- Can you give your opinion on https://github.com/matrix-org/matrix-rust-sdk/pull/3517#discussion_r1636123693 ?
- I'm currently not refreshing the sessions in memorystore with the device keys. Do I need to? Is memorystore used for any real purposes, or is it just for testing.

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (andybalaam) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs:1397]: Here is now I ran the tests locally here:

```
cd crates/matrix-sdk-indexeddb/
WASM_BINDGEN_TEST_TIMEOUT=2400 CARGO_TARGET_DIR=/tmp/andyb-rust-target wasm-pack test --firefox --headless -- --no-default-features --features e2e-encryption
```

**Review by poljar — CHANGES_REQUESTED** [review]
I think the general direction is good, left some questions/nits/suggestions.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/account.rs:1312]: Forgot to change this place to use it.

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (uhoreg) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs:924]: I don't understand what you mean by "modify the SQLite query to load both pieces of data at the same time".  Which two pieces of data are you talking about?

Alternatively, what do you think of having the cryptostore cache our own device keys, so it doesn't need to fetch it from storage each time?  It's a fairly small bit of data, so should be fine to cache.

Comment (poljar) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs:924]: > I don't understand what you mean by "modify the SQLite query to load both pieces of data at the same time". Which two pieces of data are you talking about?

Ah sorry, I meant `SQL` query. The two pieces of data are our own device keys and the session pickle. The piece of code I'm commenting on is inside the `SQLite` store. If you look at what `get_sessions_for_sender_key()` does, it executes a SQL query to fetch the serialized and possibly encrypted binary data of the Olm sessions from the database:

https://github.com/matrix-org/matrix-rust-sdk/blob/6701fd0685f59cdb86a9e11cd4aa88328ab11f49/crates/matrix-sdk-sqlite/src/crypto_store.rs#L450-L456

If you follow down the stack what `get_device()` does, you'll end up here:

https://github.com/matrix-org/matrix-rust-sdk/blob/6701fd0685f59cdb86a9e11cd4aa88328ab11f49/crates/matrix-sdk-sqlite/src/crypto_store.rs#L545-L554

So I'm wondering if we could combine those two SQL queries into one.

> Alternatively, what do you think of having the cryptostore cache our own device keys, so it doesn't need to fetch it from storage each time? It's a fairly small bit of data, so should be fine to cache.

I'm in the process of removing caches from the individual cryptostore implementations, but perhaps that will be the only way to preserve the performance. But let's do that in a separate PR if it comes to that.

Comment (uhoreg) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs]: I've tried to fetch both things in one SQL query: https://github.com/matrix-org/matrix-rust-sdk/pull/3517/commits/2d660419e5280ccc791206ce59d1b4e79d34cb9b  Is this how you were thinking of doing it?

Comment (poljar) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs]: Yeah, though benchmarking this, sadly it doesn't seem to make a difference. Let's revert it then and keep the simpler implementation. If we want our performance back we'll have to introduce a cache for our own device, though let's not bother with that in this PR.

**Review by poljar — APPROVED** [review]
Left some final nits, we should revert the SQLite query patch since it doesn't help.

Approving ahead of time, feel free to merge once the final nits have been resolved.

File: crates/matrix-sdk-sqlite/src/crypto_store.rs
Comment (uhoreg) [reply crates/matrix-sdk-sqlite/src/crypto_store.rs:1398]: I tried that, but it gave me an error saying:

```
Try find `webdriver.json` for configure browser's capabilities:
Not found
driver status: signal: 9 (SIGKILL)
```

I wrote a test anyways and it seems to pass CI, so I'll ignore it for now, and try to figure it out later if needed.

**Comment by uhoreg** [comment]
I think I've addressed all outstanding comments.  I don't have permissions to hit the merge button.  Can someone do so for me?

File: crates/matrix-sdk-crypto/src/olm/session.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/session.rs:159]: Doesn't this property need a different name:

>Until this MSC is accepted, the new property should be named org.matrix.msc4147.device_keys.

We can do this in a follow up, so we don't risk more merge conflicts.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/session.rs:159]: https://github.com/matrix-org/matrix-rust-sdk/pull/3618

**Comment by richvdh** [comment]
For some reason it doesn't seem to be mentioned in the PR :angry: but I believe this is implementing MSC4147.

**Pull request merged**
